### PR TITLE
add notes to report_conflicting_borrow MIR borrowck

### PR DESCRIPTION
**Pull request by mikhail-m1**
part of #44596

### CODE REVIEW

File: src/librustc_mir/borrow_check.rs [hunk 1]
```diff
@@ -934,43 +932,53 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
...
-        let (loan2_lvalue, loan2_kind) = loan2;
+                                 gen_borrow_kind: BorrowKind,
+                                 issued_borrow: &BorrowData) {
         // FIXME: obviously falsifiable. Generalize for non-eq lvalues later.
-        assert_eq!(loan1_lvalue, loan2_lvalue);
+        assert_eq!(lvalue, &issued_borrow.lvalue);
+
+        let issued_span = self.retrieve_borrow_span(issued_borrow);
+
+        let issued_end_loan_span = if let Lvalue::Local(local) = issued_borrow.lvalue {
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: That's not how you do it. You should look up the span of the `EndRegion`.

File: src/librustc_borrowck/borrowck/check_loans.rs [hunk 2]
```diff
@@ -470,102 +470,29 @@ impl<'a, 'tcx> CheckLoanCtxt<'a, 'tcx> {
...
-                            "borrow ends here");
-                    err
-                }
-
+                (ty::MutBorrow, ty::MutBorrow) =>
+                    self.bccx.cannot_mutably_borrow_multiply(
+                        new_loan.span, &nl, &new_loan_msg, old_loan.span, &old_loan_msg,
+                        previous_end_span, Origin::Ast),
+                (ty::UniqueImmBorrow, ty::UniqueImmBorrow) =>
+                    self.bccx.cannot_uniquely_borrow_by_two_closures(
```
Comment (arielb1) [inline src/librustc_borrowck/borrowck/check_loans.rs]: I'm not sure how "unique" became related to closures. I mean, unique implies closure but not vice-versa (closures can also borrow immutably). But that's pre-existing.

File: src/librustc_mir/borrow_check.rs [hunk 3]
```diff
@@ -934,43 +932,53 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
...
                                  (lvalue, span): (&Lvalue, Span),
-                                 loan1: (&Lvalue, BorrowKind),
-                                 loan2: (&Lvalue, BorrowKind)) {
-        let (loan1_lvalue, loan1_kind) = loan1;
-        let (loan2_lvalue, loan2_kind) = loan2;
+                                 gen_borrow_kind: BorrowKind,
+                                 issued_borrow: &BorrowData) {
         // FIXME: obviously falsifiable. Generalize for non-eq lvalues later.
-        assert_eq!(loan1_lvalue, loan2_lvalue);
+        assert_eq!(lvalue, &issued_borrow.lvalue);
```
Comment (arielb1) [inline src/librustc_mir/borrow_check.rs]: Could you add a message to this assertions, e.g.:
```Rust
assert_eq!(lvalue, &issued_borrow.lvalue, "complex conflicts are not supported");
```

**Review by arielb1 — CHANGES_REQUESTED** [review]
generally LGTM, but you should get the region-end span the right way - I think you should create a Region -> span-of-endregion map by indexing the MIR, and look the region up in that.

File: src/librustc_mir/borrow_check.rs (hunk 1 above)
Comment (mikhail-m1) [reply src/librustc_mir/borrow_check.rs]: I've spent some time trying to do this, but `StatementKind::EndRegion` contains `region::Scope`, and I cannot get span from it. ```
impl Scope {
 pub fn span(&self, tcx: TyCtxt, scope_tree: &ScopeTree)
```
Where should I get ScopeTree? I found MIR build already do this and resided to use it result.

Comment (pnkfelix) [reply src/librustc_mir/borrow_check.rs]: @mikhail-m1 I think the idea is that the `EndRegion` *itself* is a `Statement`, and every `Statement` carries a `SourceInfo` which in turn carries a `Span`.

File: src/librustc_mir/dataflow/impls/borrows.rs [hunk 4]
```diff
@@ -90,6 +97,19 @@ impl<'a, 'tcx> Borrows<'a, 'tcx> {
                     borrows.insert(idx);
                 }
             }
+
+            fn visit_statement(&mut self,
+                               block: mir::BasicBlock,
+                               statement: &mir::Statement<'tcx>,
+                               location: Location) {
+                if let mir::StatementKind::EndRegion(region_scope) = statement.kind {
+                    error!("statement {:?} {:?}", statement, statement.source_info.span);
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/impls/borrows.rs]: Is this supposed to be in here? Or was this instrumentation you put in but was supposed to be removed before merging?

File: src/librustc_mir/dataflow/impls/borrows.rs [hunk 5]
```diff
@@ -90,6 +97,19 @@ impl<'a, 'tcx> Borrows<'a, 'tcx> {
...
+                    if let Some(span) = self.region_span_map.insert(ReScope(region_scope), statement.source_info.span) {
```
Comment (pnkfelix) [inline src/librustc_mir/dataflow/impls/borrows.rs]: I think the tidy pass is complaining about this line being too long. (We have a hard 100-character limit on line length.)

File: src/librustc_mir/borrow_check.rs [hunk 6]
```diff
@@ -396,26 +396,31 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
                             ReadKind::Copy =>
                                 this.report_use_while_mutably_borrowed(
                                     context, lvalue_span, borrow),
-                            ReadKind::Borrow(bk) =>
+                            ReadKind::Borrow(bk) =>{
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check.rs]: nit: coding style in project is to put a single space between the `=>` and the `{` in a match arm

File: src/librustc_mir/borrow_check.rs [hunk 7]
```diff
@@ -396,26 +396,31 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
...
+                                let end_issued_loan_span =
+                                    flow_state.borrows.base_results.operator().region_span(
+                                        &borrow.region).end_point();
                                 this.report_conflicting_borrow(
-                                    context, lvalue_span,
-                                    (lvalue_span.0, bk), (&borrow.lvalue, borrow.kind)),
+                                    context, lvalue_span, bk, &borrow, end_issued_loan_span)
+                            },
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check.rs]: nit: coding style in project is to leave out the trailing comma when a match arm's code is delimited by curly braces

File: src/librustc_mir/borrow_check.rs [hunk 8]
```diff
@@ -396,26 +396,31 @@ impl<'c, 'b, 'a: 'b+'c, 'gcx, 'tcx: 'a> MirBorrowckCtxt<'c, 'b, 'a, 'gcx, 'tcx>
...
                         match kind {
-                            WriteKind::MutableBorrow(bk) =>
+                            WriteKind::MutableBorrow(bk) => {
+                                let end_issued_loan_span =
+                                    flow_state.borrows.base_results.operator().region_span(
+                                        &borrow.region).end_point();
                                 this.report_conflicting_borrow(
-                                    context, lvalue_span,
-                                    (lvalue_span.0, bk), (&borrow.lvalue, borrow.kind)),
+                                    context, lvalue_span, bk, &borrow, end_issued_loan_span)},
```
Comment (pnkfelix) [inline src/librustc_mir/borrow_check.rs]: nit: coding style in project is to both omit the trailing comma here (see earlier comment) and also put a line break before the ending `}`

**Comment by mikhail-m1** [comment]
look like some bug with powerpc64 build
ERROR:  Could not find a valid gem 'aws-sdk-resources' (= 2.10.57) in any repository

**Pull request merged**